source/Cosmos.System2/Network/IPv4/TCP/TCPClient.cs

```diff
@@ -0,0 +1,475 @@
+
+                SequenceNumber++;
+
+                Status = Status.FIN_WAIT1;
+
+                if (WaitStatus(Status.CLOSED, 5000) == false)
+                {
+                    throw new Exception("Failed to close TCP connection!");
+                }
+            }
+
+            if (clients.ContainsKey((uint)localPort) == true)
+            {
+                clients.Remove((uint)localPort);
+            }
+        }
+
+        /// <summary>
+        /// Send data to client.
+        /// </summary>
+        /// <param name="data">Data array to send.</param>
+        /// <exception cref="Exception">Thrown if destination is null or destinationPort is 0.</exception>
+        /// <exception cref="ArgumentException">Thrown on fatal error (contact support).</exception>
+        /// <exception cref="OverflowException">Thrown if data array length is greater than Int32.MaxValue.</exception>
+        /// <exception cref="Sys.IO.IOException">Thrown on IO error.</exception>
+        public void Send(byte[] data)
+        {
+            if ((destination == null) || (destinationPort == 0))
+            {
+                throw new InvalidOperationException("Must establish a default remote host by calling Connect() before using this Send() overload");
+            }
+            if (data.Length > 1500)
+            {
+                throw new NotImplementedException("Data length must be less than 1500 bytes (yet!)");
+            }
+
+            var packet = new TCPPacket(source, destination, (ushort)localPort, (ushort)destinationPort, SequenceNumber, AckNumber, 20, 0x18, 0xFAF0, 0, data);
+            OutgoingBuffer.AddPacket(packet);
+            NetworkStack.Update();
+
+            SequenceNumber += (uint)data.Length;
+
+            /*if (WaitStatus(Status.OPENED, 5000) == false)
+            {
+                throw new Exception("Failed to send TCP data!");
+            }*/
+        }
+
+        /// <summary>
+        /// Receive data from end point.
+        /// </summary>
+        /// <param name="source">Source end point.</param>
+        /// <returns>byte array value.</returns>
+        /// <exception cref="InvalidOperationException">Thrown on fatal error (contact support).</exception>
+        public byte[] NonBlockingReceive(ref EndPoint source)
+        {
+            if (rxBuffer.Count < 1)
+            {
+                return null;
+            }
+
+            var packet = new TCPPacket(rxBuffer.Dequeue().RawData);
+            source.address = packet.SourceIP;
+            source.port = packet.SourcePort;
+
+            return packet.TCP_Data;
+        }
+
+        /// <summary>
+        /// Receive data from end point.
+        /// </summary>
+        /// <param name="source">Source end point.</param>
+        /// <returns>byte array value.</returns>
+        /// <exception cref="InvalidOperationException">Thrown on fatal error (contact support).</exception>
+        public byte[] Receive(ref EndPoint source)
```

Comment:
Can we also specify a timeout to the wait?

Reply:
I think this should be managed at the user level, like UDP